Changed functionality when a user sends a new message from a post page.

jcr7467 opened the pull request:
Now, instead of first redirecting them to the chat page, we send a message from that same page. This saves us a lot of trouble when sending a new message from the chat page.

We also hide the chat buttom that sends us to the chat page from navbars until a user has at least 1 person they are talking to.

jcr7467 commented:
Couples of notes:

Because we now send a new message from the actual post page, I removed the if statement in our JS handlebars partial that checked if the 'bookOwner' query object was passed in from response.render. 
So this entire file I deleted and put the other remaining file that remains relevant back in chat.js

Along with another issue I am working on (CSS equivalent) I added to the JS handlebars partial a script includer that is a variable 'filename' so that we can pass in JS for a specific page.
This is so that we can trim back things from our styles.js file and decrease load speed by not importing unneeded JS files

ntjoar reviewed (changes requested):
So on testing this, I noticed a few things.

First, we have unverified users able to send messages, but not view their chats. Even after they send a message, there is no chat icon. Login info for this account example can be given on request. 

Second, we are able to send null and empty messages. This should be a relatively simple fix on checking that text_inputted.trim() is not empty and is not null.

Finally, upon using my own account to check messages, I am getting a book owner is not defined error for some reason. Maybe it's part of the changes you commented on, but I could not see why I was getting this error.

jcr7467 commented:
Brilliant catches, thank you. Will work bugs and report back when done.

jcr7467 commented:
Fixed sending null/white space messages

jcr7467 commented:
Fixed error for checking for bookOwner

jcr7467 commented:
The issue of dynamically displaying the chat button remains.

I managed to get the button to appear after sending a first message, but after navigating away from the page, it disappears again.

Normally, the solution to our problem would be to set the currentUserObject, and on reload it would apply(shown below), but it is not working in this case. I suspect that it is because of how our chat system works. 

Because we use sockets to first send a message, and then on the server, we do a seperate AJAX request to another route, I suspect that it is not applying directly as we would hope.
To test that the DB object was indeed updated:
1.  I made a new account, and sent a message.
2. Witnessed the chat button show up on screen after modal closed
3. Navigated to another page, and saw that it again disappeared, suggesting currentUserObject was not updated
4. Logged out and logged back in, and saw that the chat icon did indeed show up.

<img width="642" alt="Screen Shot 2020-11-12 at 11 51 50 PM" src="https://user-images.githubusercontent.com/25701682/99042899-61929e00-2542-11eb-976f-525fa4a195c5.png">

jcr7467 commented:
Also, as I was working on this, a thought came across. If two users are logged in at the same time, both new users with no messages, and one of them sends the other one a new message. 

How would the recipient's session know that it received a message? 
If we did not think about this, the recipient would not have their chat button appear until they logged out and logged back in or if they themselves made some update to their account, despite the fact that they have messages to see.

We would have to have some sort of listener that is constantly checking for changes to the DB object of the user. This seems expensive to implement.

An easy solution would be to not hide the messages button ever, and have a more visually pleasing page when a user has no messages.

jcr7467 commented:
Removed the feature of the hidden chat message, for now. To try to make up for it, I added if statements in conversation_list.hbs that displays a certain message if a user has no messages, or if a user has not verified their email.

Let me know on thoughts & opinions on this fix. (:

jcr7467 commented on routes/message_routes.js:
```diff
@@ -291,31 +292,12 @@ router.route('/conversations').get((request, response, next) => {
 
                 }).then(([userMap, penpalCount]) => {
 
-                    if (bookOwner){
```
Looks a bit confusing, but just removed any instance that checked for 'bookOwner'. We only needed this because if that variable was passed in, then it created a new message to a new user, but because we changed that implementation, we no longer need this.

jcr7467 commented on views/bookpage.hbs:
```diff
@@ -264,6 +258,60 @@
 
 
 
```
Below this comment is the new modal we use to send a new message

jcr7467 commented on views/conversation_list.hbs:
```diff
@@ -58,25 +58,59 @@
             <!--    </div>-->
```
My attempt as making up for difficulty in hiding/showing messages button. If statements that check for messages, and for if a user is verified yet.

jcr7467 commented on views/partials/navbars/search-header-index.hbs:
```diff
@@ -68,17 +68,20 @@
                 <div class="col-6 col-sm-6 col-md-4 col-lg-2">
```
Removed the check for conversation length from all 3 navbars, so those disappeared from this changelog, this is this is the only one that actually had changes.

ntjoar reviewed (approved):
Looks good. I think in regards to getting new messages, we may have to make a backend call and add a variable of seen or not seen for the chat message. We can probably look into this for further investigation

jcr7467 commented:
> Looks good. I think in regards to getting new messages, we may have to make a backend call and add a variable of seen or not seen for the chat message. We can probably look into this for further investigation

Should we make this a new issue or moreso just add it to the notification system issue

jcr7467 reviewed (commented):
One last review for myself lol: looks good to me
